k_shot_testing.py

```diff
@@ -0,0 +1,133 @@
+import os
+
+from transformers import BertTokenizer, AdamW
+
+from util import get_test_args, get_pytorch_device, load_model
+from tasks import *
+from models import MultiTaskLearner, PrototypeLearner, ProtoMAMLLearner
+
+import torch.optim as optim
+import torch.nn as nn
+
+
+def k_shot_testing(model, episodes, test_task, device, num_updates=5, num_test_batches=None, lr=5e-5, zero_init=False):
+    # get iterator over test task
+    tokenizer = BertTokenizer.from_pretrained('bert-base-uncased', do_lower_case=True)
+    test_iter = test_task.get_iter('test', tokenizer, shuffle=False)
+
+    # set number of test batches to evaluate on
+    test_size = len(test_iter)
+    if num_test_batches is None or num_test_batches > test_size:
+        num_test_batches = test_size
+
+    # Define optimizers and loss function
+    optimizer = optim.SGD(params=model.parameters(), lr=lr)
+    cross_entropy = nn.CrossEntropyLoss()
+
+    model.to(device)
+
+    episode_accs = []
+    for episode in episodes:
+        model.train()
+        # setup output layer for ProtoMAML and MTL
+        if isinstance(model, ProtoMAMLLearner):
+            proto_embeddings = model.proto_net(episode[0].to(device), attention_mask=episode[2].to(device))
+            prototypes = model.proto_net.calculate_centroids((proto_embeddings, episode[1]), test_task.num_classes)
+            W, b = model.calculate_output_params(prototypes.detach())
+            model.initialize_classifier(W, b)
+        elif zero_init and isinstance(model, MultiTaskLearner):
+            task_module_name = 'task_{}'.format(test_task.get_name())
+            out_MTL_layer = model._modules[task_module_name]
+            out_MTL_layer.weight.data = torch.zeros_like(out_MTL_layer.weight.data)
+            out_MTL_layer.bias.data = torch.zeros_like(out_MTL_layer.bias.data)
+
+        # fine-tune model with some updates on the provided episode
+        for update in range(num_updates):
```

Comment:
Should we move this loop outside the episodes one. It seems we might be going full force towards solving this episode by using it `num_updates` times, this would be the same as to increasing the learning rate and doing only one pass, no?